crates/bindings-csharp/BSATN.Codegen/Diag.cs

```diff
@@ -0,0 +1,162 @@
+        Func<TContext, ISymbol> toLocation
+    )
+        : this(group, title, interpolate, ctx => toLocation(ctx).Locations.FirstOrDefault()) { }
+
+    public Diagnostic ToDiag(TContext ctx) =>
+        Diagnostic.Create(descriptor, toLocation(ctx), makeFormatArgs(ctx));
+}
+
+internal static class ErrorDescriptor
+{
+    private static readonly ErrorDescriptorGroup group = new("BSATN", "SpacetimeDB.BSATN");
+
+    public static readonly ErrorDescriptor<(
+        ISymbol member,
+        ITypeSymbol type,
+        Exception e
+    )> UnsupportedType =
+        new(
+            group,
+            "Unsupported type",
+            ctx => $"BSATN implementation for {ctx.type} is not found: {ctx.e.Message}",
+            ctx => ctx.member
+        );
+
+    public static readonly ErrorDescriptor<(
+        EqualsValueClauseSyntax equalsValue,
+        EnumMemberDeclarationSyntax enumMember,
+        EnumDeclarationSyntax @enum
+    )> EnumWithExplicitValues =
+        new(
+            group,
+            "[SpacetimeDB.Type] enums cannot have explicit values",
+            ctx =>
+                $"{ctx.@enum.Identifier}.{ctx.enumMember.Identifier} has an explicit value {ctx.equalsValue.Value} which is not allowed in SpacetimeDB enums.",
+            ctx => ctx.equalsValue
+        );
+
+    public static readonly ErrorDescriptor<EnumDeclarationSyntax> EnumTooManyVariants =
+        new(
+            group,
+            "[SpacetimeDB.Type] enums are limited to 256 variants",
+            @enum =>
+                $"{@enum.Identifier} has {@enum.Members.Count} variants which is more than the allowed 256 variants for SpacetimeDB enums.",
+            @enum => @enum.Members[256]
+        );
+
+    public static readonly ErrorDescriptor<INamedTypeSymbol> TaggedEnumInlineTuple =
+        new(
+            group,
+            "Tagged enum variants must be declared with inline tuples",
+            baseType =>
+                $"{baseType} does not have the expected format SpacetimeDB.TaggedEnum<(TVariant1 v1, ..., TVariantN vN)>.",
+            baseType => baseType
+        );
+
+    public static readonly ErrorDescriptor<IFieldSymbol> TaggedEnumField =
+        new(
+            group,
+            "Tagged enums cannot have instance fields",
+            field =>
+                $"{field.Name} is an instance field, which are not permitted inside SpacetimeDB tagged enums.",
+            field => field
+        );
+
+    public static readonly ErrorDescriptor<TypeParameterListSyntax> TypeParams =
+        new(
+            group,
+            "Type parameters are not yet supported",
+            typeParams => $"Type parameters {typeParams} are not supported in SpacetimeDB types.",
+            typeParams => typeParams
+        );
+}
+
+public interface IDiagHolder
+{
+    Diagnostics Diag { get; }
+}
+
+public class Diagnostics : EquatableList<Diagnostic>, IDiagHolder
+{
+    Diagnostics IDiagHolder.Diag => this;
+}
```

Comment:
Slightly confused about this setup, what exactly is going on here?

Would appreciate a comment on IDiagHolder saying "should be implemented by ..., make sure you call Utils.ReportDiagnostics after you ..."

Reply:
Oh this is just a convenience for `ReportDiagnostics` to handle anything that has `Diagnostics Diag;` inside, as well as raw `Diagnostics` object itself (the last one is used in enum checks).

I'll add comments.